proposals/0025/README.md

```diff
@@ -0,0 +1,199 @@
+[watch]: https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes
+[fm]: https://kubernetes.io/docs/reference/using-api/server-side-apply/#field-management
+[capi]: https://cluster-api.sigs.k8s.io/
+[capt]: https://github.com/tinkerbell/cluster-api-provider-tink
+
+## Content
+
+### Architectural Motivation
+
+Tinkerbell is very flexible and uses standard protocols like PXE, DHCP, and HTTP to support provisioning hardware.
+In order to support more features, more API machinery work will be required.
+Listed below are some of the contributing architectural motivations.
+
+* Streaming updates to clients is difficult with the current database
+* Multiple-worker workflows will require numerous API changes
+* There is no authorization in the Tinkerbell API today.
+  In order to support a least-privilege access model, several new authorization modes would need to be supported
+  * An attribute-based authorization method for Tink workers.
+    At the time of writing, any Tink worker is authorized to list all hardware data, including possibly sensitive metadata for other hardware.
+    Today, the Tinkerbell Cluster API (CAPI) provider stores machine CA private key data in Tinkerbell hardware metadata.
+    Ideally, some authorizer that could make decisions based on authentication identity would be used (The Kubernetes analog would be the [Node Authorizer][node-authorizer]).
+    The authenticatoin method in this model could be a network identity, a TPM, or some other bootstrapped authentication credential.
+  * A role-based access method administrative clients.
+  * Either a role-based access method, or some forwarding mechanism for passthrough clients like Boots and Hegel.
+* Operating high-availability Postgres is non-trivial in many environments.
+
+[node-authorizer]: https://kubernetes.io/docs/reference/access-authn-authz/node/
+
+If we only focus on the Tinkerbell API clients as it exists in Q3 2021, the architecture looks something like the following:
+
+[![current-architecture](./tink-arch-1.png)](.tink-arch-2.png)
+
+In this proposal, all non-worker clients of Tinkerbell would become Kubernetes clients.
+
+[![proposed-architecture](./tink-arch-2.png)](./tink-arch-2.png)
+
+By using the Kubernetes API as the datastore, the motivations are addressed in the following ways:
+
+* Kubernetes natively supports streaming watches for CRD types.
+  Any process (PBnJ, Tink API, CAPT, etc) that need streaming updates can leverage Kubernetes Watches.
+* Multi-worker workflows will require some backend changes, but that will be opaque to Tink worker clients.
+* The only authorization mode required for the Tinkerbell API will be to support
+  * All Kubernetes clients (Boots, Hegel, PBnJ, etc) can have least-permission RBAC policies to limit the permissions for each respective client.
+* High-availability Kubernetes and etcd can be delegated to a cloud provider or an on-premise Kubernetes cluster
+
+### Tradeoffs
+
+TODO
```

Comment:
You may want to mention something about the bootstrapping/circular dependency issue here, as often the nodes of a Kubernetes cluster would be provisioned by a system like Tinkerbell.

The two cases to think about in particular are initial setup and emergency provisioning during an outage.

Reply:
I added a section on Administration and addressed bootstrapping and backup/restore.